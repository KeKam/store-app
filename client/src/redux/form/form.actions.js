import FormActionTypes from './form.types';

export const startSendForm = (contactInformation) => ({
  type: FormActionTypes.START_SEND_FORM,
  payload: contactInformation,
});

export const sendFormSuccess = (form) => ({
  type: FormActionTypes.SEND_FORM_SUCCESS,
  payload: form,
});

export const sendFormFailure = (error) => ({
  type: FormActionTypes.SEND_FORM_FAILURE,
  payload: error,
});

export const resetFormAfterSuccess = () => ({
  type: FormActionTypes.RESET_FORM_AFTER_SUCCESS,
});
