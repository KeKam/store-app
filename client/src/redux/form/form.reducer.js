import FormActionTypes from './form.types';

const DEFAULT_STATE = {
  formWasSent: null,
  isSending: false,
  error: null
};

const formReducer = (state = DEFAULT_STATE, action) => {
  switch (action.type) {
    case FormActionTypes.START_SEND_FORM:
      return {
        ...state,
        isSending: true,
        error: null
      };
    case FormActionTypes.SEND_FORM_SUCCESS:
      return {
        ...state,
        formWasSent: action.payload,
        isSending: false,
        error: null
      };
    case FormActionTypes.SEND_FORM_FAILURE:
      return {
        ...state,
        error: action.payload,
        isSending: false
      };
    case FormActionTypes.RESET_FORM_AFTER_SUCCESS:
      return DEFAULT_STATE;
    default:
      return state;
  }
};

export default formReducer;
