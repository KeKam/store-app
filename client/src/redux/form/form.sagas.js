import { takeLatest, put, call, all } from 'redux-saga/effects';

import FormActionTypes from './form.types';
import { saveMessage } from '../../firebase/firebase';
import {
  sendFormSuccess,
  sendFormFailure,
  resetFormAfterSuccess,
} from './form.actions';

export const delay = (ms) => new Promise((res) => setTimeout(res, ms));

export function* sendForm({ payload: { name, email, description } }) {
  try {
    yield saveMessage(name, email, description);
    yield put(sendFormSuccess({ name, email, description }));
  } catch (error) {
    yield put(sendFormFailure(error));
  }
}

export function* resetForm() {
  yield delay(3000);
  yield put(resetFormAfterSuccess());
}

export function* onStartSendForm() {
  yield takeLatest(FormActionTypes.START_SEND_FORM, sendForm);
}

export function* onSendFormSuccess() {
  yield takeLatest(FormActionTypes.SEND_FORM_SUCCESS, resetForm);
}

export function* formSagas() {
  yield all([call(onStartSendForm), call(onSendFormSuccess)]);
}
