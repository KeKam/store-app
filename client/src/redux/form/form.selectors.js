import { createSelector } from 'reselect';

const selectForm = state => state.form;

export const selectIsSendingForm = createSelector(
  [selectForm],
  form => form.isSending
);

export const selectFormError = createSelector([selectForm], form => form.error);

export const selectFormWasSent = createSelector(
  [selectForm],
  form => form.formWasSent
);
