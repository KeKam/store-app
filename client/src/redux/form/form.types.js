const FormActionTypes = {
  START_SEND_FORM: 'START_SEND_FORM',
  SEND_FORM_SUCCESS: 'SEND_FORM_SUCCESS',
  SEND_FORM_FAILURE: 'SEND_FORM_FAILURE',
  RESET_FORM_AFTER_SUCCESS: 'RESET_FORM_AFTER_SUCCESS',
};

export default FormActionTypes;
