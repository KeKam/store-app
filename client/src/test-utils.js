import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';

import rootReducer from './redux/root-reducer';

// const render = (
//   component,
//   {
//     initialState,
//     store = createStore(rootReducer, initialState),
//     ...renderOptions
//   } = {}
// ) => {
//   return {
//     ...rtlRender(<Provider store={store}>{component}</Provider>),
//     store,
//   };
//   // const Wrapper = ({ children }) => {
//   //   return <Provider store={store}>{children}</Provider>;
//   // };
//   // return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// };

const render = (component, { initialState } = {}) => {
  const actions = [];
  const observerMiddleware = () => (next) => (action) => {
    actions.push(action);
    return next(action);
  };
  const store = createStore(
    rootReducer,
    initialState,
    applyMiddleware(observerMiddleware)
  );
  const utils = {
    dispatch(action) {
      return store.dispatch(action);
    },
    getDispatchedActions() {
      return actions;
    },
    getState() {
      return store.getState();
    },
  };
  return {
    ...rtlRender(<Provider store={store}>{component}</Provider>),
    ...utils,
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
