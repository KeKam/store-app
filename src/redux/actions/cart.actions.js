import { CartActionTypes } from '../types/cart.types';

export const toggleCart = () => {
  return {
    type: CartActionTypes.TOGGLE_CART
  };
};

export const addToCart = item => {
  return {
    type: CartActionTypes.ADD_TO_CART,
    payload: item
  };
};

export const removeFromCart = item => {
  return {
    type: CartActionTypes.REMOVE_FROM_CART,
    payload: item
  };
};

export const removeItem = item => {
  return {
    type: CartActionTypes.REMOVE_ITEM,
    payload: item
  };
};
