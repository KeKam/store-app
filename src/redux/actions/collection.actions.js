import { CollectionActionTypes } from '../types/collection.types';
import {
  firestore,
  convertCollectionSnapshotToMap
} from '../../firebase/firebase';

export const startFetchCollection = () => ({
  type: CollectionActionTypes.START_FETCH_COLLECTION
});

export const fetchCollectionSuccess = collectionMap => ({
  type: CollectionActionTypes.FETCH_COLLECTION_SUCCESS,
  payload: collectionMap
});

export const fetchCollectionFailure = error => ({
  type: CollectionActionTypes.FETCH_COLLECTION_FAILURE,
  payload: error
});

export const startAsyncFetchCollection = () => {
  return async dispatch => {
    const collectionRef = firestore.collection('collection');
    dispatch(startFetchCollection());

    try {
      const response = await collectionRef.get();
      const collectionMap = convertCollectionSnapshotToMap(response);
      dispatch(fetchCollectionSuccess(collectionMap));
    } catch (error) {
      dispatch(fetchCollectionFailure(error));
    }
  };
};
