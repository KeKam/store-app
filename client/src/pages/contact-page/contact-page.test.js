import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { render, screen, fireEvent } from '../../test-utils';
import { startSendForm } from '../../redux/form/form.actions';
import ContactPage from './contact-page';

describe('ContactPage component', () => {
  it('should render ContactPage component', () => {
    const { asFragment } = render(<ContactPage />);

    expect(asFragment()).toMatchSnapshot();
    expect(
      screen.getByRole('heading', { name: 'Need to contact us?' })
    ).toBeInTheDocument();
    expect(screen.getByText('Please use the form below')).toBeInTheDocument();
    expect(screen.getByTestId('name-input')).toHaveDisplayValue('');
    expect(screen.getByTestId('email-input')).toHaveDisplayValue('');
    expect(screen.getByTestId('description-input')).toHaveDisplayValue('');
    expect(screen.getByRole('button', { name: 'SEND' })).toBeInTheDocument();
    expect(screen.queryByTestId('spinner')).not.toBeInTheDocument();
    expect(
      screen.queryByText('Something went wrong, please try again')
    ).not.toBeInTheDocument();
  });

  it('should change display value for name on change', () => {
    render(<ContactPage />);

    userEvent.type(screen.getByTestId('name-input'), 'TestUser');

    expect(screen.getByTestId('name-input')).toHaveDisplayValue('TestUser');
  });

  it('should change display value for email on change', () => {
    render(<ContactPage />);

    userEvent.type(screen.getByTestId('email-input'), 'testUser@gmail.com');

    expect(screen.getByTestId('email-input')).toHaveDisplayValue(
      'testUser@gmail.com'
    );
  });

  it('should change display value for description on change', () => {
    render(<ContactPage />);

    fireEvent.change(screen.getByTestId('description-input'), {
      target: { value: 'Need help' },
    });

    expect(screen.getByTestId('description-input')).toHaveDisplayValue(
      'Need help'
    );
  });

  it('should dispatch startSendForm action on click', () => {
    const mockState = {
      name: '',
      email: '',
      description: '',
    };

    const { getDispatchedActions } = render(<ContactPage />);

    userEvent.click(screen.getByRole('button', { name: 'SEND' }));

    expect(getDispatchedActions()).toContainEqual(startSendForm(mockState));
  });

  it('should render Spinner component', () => {
    const mockState = {
      formWasSent: null,
      isSending: true,
      error: null,
    };

    render(<ContactPage />, { initialState: { form: mockState } });

    expect(screen.getByTestId('spinner')).toBeInTheDocument();
    expect(screen.queryByText('Need to contact us?')).not.toBeInTheDocument();
  });

  it('should render confirmation message', () => {
    const mockState = {
      formWasSent: {
        name: 'TestUser',
        email: 'testuser@gmail.com',
        description: 'I need to test',
      },
      isSending: false,
      error: null,
    };

    render(<ContactPage />, { initialState: { form: mockState } });

    expect(screen.getByText('Message sent')).toBeInTheDocument();
    expect(
      screen.getByText('We will get back to you soon')
    ).toBeInTheDocument();
    expect(screen.queryByText('Need to contact us?')).not.toBeInTheDocument();
    expect(screen.queryByTestId('spinner')).not.toBeInTheDocument();
  });

  it('should render error message', () => {
    const mockState = {
      formWasSent: null,
      isSending: false,
      error: 'Something wrong',
    };

    render(<ContactPage />, { initialState: { form: mockState } });

    expect(
      screen.getByText('Something went wrong, please try again')
    ).toBeInTheDocument();
  });
});
