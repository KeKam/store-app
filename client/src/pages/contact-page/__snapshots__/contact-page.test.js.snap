// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactPage component should render ContactPage component 1`] = `
<DocumentFragment>
  <div
    class="sc-bZQynM kzvvTl"
  >
    <div
      class="sc-gzVnrw gytjGv"
    >
      <h2
        class="sc-htoDjs GlcDy"
      >
        Need to contact us?
      </h2>
      <span>
        Please use the form below
      </span>
      <p
        class="sc-dnqmqq eLXxaK"
      />
      <form>
        <div
          class="sc-bdVaJa eEzrUY"
        >
          <input
            class="sc-bwzfXH eKtSpl"
            data-testid="name-input"
            name="name"
            required=""
            type="text"
            value=""
          />
          <label
            class="sc-htpNat gLmdTI"
            value=""
          >
            Name
          </label>
        </div>
        <div
          class="sc-bdVaJa eEzrUY"
        >
          <input
            class="sc-bwzfXH eKtSpl"
            data-testid="email-input"
            name="email"
            required=""
            type="email"
            value=""
          />
          <label
            class="sc-htpNat gLmdTI"
            value=""
          >
            Email
          </label>
        </div>
        <div
          class="sc-bdVaJa eEzrUY"
        >
          <textarea
            class="sc-bwzfXH eKtSpl"
            data-testid="description-input"
            name="description"
            required=""
            rows="5"
            type="text"
          />
          <label
            class="sc-htpNat gLmdTI"
            value=""
          >
            Description
          </label>
        </div>
        <button
          class="sc-bxivhb jzXiqt"
          type="submit"
        >
           SEND 
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
